### PR TITLE
Collect checks to run in CrossVersionChecks

### LINKED ISSUE #2968: print deprecation warning for bitshift Int by a Long
The current situation is pretty bad:

2.11/Dotty:
```scala
scala> 1 << 33L
res0: Long = 8589934592

scala> val x = 1
x: Int = 1

scala> x << 33L
res1: Int = 0
```

2.12 (since https://github.com/scala/scala/pull/4238):
```scala
scala> 1 << 33L
res0: Int = 2

scala> val x = 1
x: Int = 1

scala> x << 33L
res1: Int = 2
```

See also the discussion at https://gitter.im/scala/contributors?at=598b4b3eee5c9a4c5fa61568 where @ichoran proposed getting rid of `Int#<<(x: Long)` since it can be so easily misused.

### PR DESCRIPTION
**Pull request by som-snytt**
Check for deprecation when inlining or constant folding, but avoid cyclic errors by deferring the exclusion check, which walks the owner chain, until CrossVersionChecks, where the check normally happens.

Fixes #2968

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/core/Contexts.scala [hunk 1]
```diff
@@ -52,8 +52,10 @@ object Contexts {
   private val (notNullInfosLoc,     store8) = store7.newLocation[List[NotNullInfo]]()
   private val (importInfoLoc,       store9) = store8.newLocation[ImportInfo | Null]()
   private val (typeAssignerLoc,    store10) = store9.newLocation[TypeAssigner](TypeAssigner)
+  private val (usagesLoc,          store11) = store10.newLocation[Any]()          // unusages feature
+  private val (deferredChecksLoc,  store12) = store11.newLocation[DeferredChecks]()
 
-  private val initialStore = store10
+  private val initialStore = store12
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/core/Contexts.scala:58]: Adding something to the store is an expensive operation. It charges every time we copy a store with two additional slots to copy. if we look up the property not very often it's better to use a Property.

**Comment by odersky** [comment]
Is it reallly #2968 that is fixed here?

**Comment by som-snytt** [comment]
Scala 2 has a similar but simpler mechanism just to run functions after typer, with the same goal of avoiding cycles.

This could just be tacked onto `CrossVersionChecks`, more tightly coupled but also sufficient.

**Comment by odersky** [comment]
Yes, why don't we do this in cross version checks? That's where the other tests are done, right?

**Comment by som-snytt** [comment]
Constant folded ops currently disappear. Instead of trees preserved and constant types assigned.

**Comment by odersky** [comment]
What is an op that is constant folded and deprecated? Is there anything that's realistic in practice?

**Comment by odersky** [comment]
The other possibility is to order phases differently. We can move CrossVersionChecks earlier (with a bit of massaging) or drop trees of constant types later (currently done in FirstTransform).

**Pull request closed without merging**